### PR TITLE
[CDAP-6661] Order by aggregate plugin

### PR DESCRIPTION
**Pull request by ananya-bhattacharya**
The Order By a aggregator plugin that sorts structured records based on fields and order specified by user. 
For design and use case please refer - https://wiki.cask.co/display/CE/Order+By

### CODE REVIEW

File: orderby-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+import java.lang.reflect.Type;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+/**
+ * CustomWritable for the composite key.
+ */
+public class CompositeKey implements Writable, WritableComparable<CompositeKey> {
+
+  private static final String FIELD_DELIMITER = "->";
```
Comment (nitinmotgi) [inline orderby-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java]: Why such a delimiter ? Can't we use something standard ?

File: orderby-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+  private final Type mapType = new TypeToken<LinkedHashMap>() {  }.getType();
+  private final Type schemaType = new TypeToken<Schema>() {  }.getType();
+  private String structureRecordJSON;
+  private String sortFieldsJSON;
+  private String schemaJSON;
+
+  public CompositeKey() {
+  }
+
+  public CompositeKey(Text structureRecordJSON, String sortFieldsJSON, String schemaJSON) {
```
Comment (nitinmotgi) [inline orderby-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java]: structuredRecord not structureRecord

File: orderby-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+  /**
+   * Get sorting order for structured records depending on the type of value contained in key
+   *
+   * @param key               contains the value to sort the data
+   * @param value             determines whether the objects are to be sorted in ascending or descending order for key
+   * @param structuredRecord1 Structured record
+   * @param structuredRecord2 Structured record to compare against
+   * @return sorting order of the complete structured record
+   */
+  public static int compareRecords(String key, String value, StructuredRecord structuredRecord1,
```
Comment (nitinmotgi) [inline orderby-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java:80]: Comparison has to support recursive complex records. So you might want to implement a recursive function for doing that. StructuredRecord can have basic types as well as Record, Union, etc.

File: orderby-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+      res = getSortingOrder(value) * Byte.compare((Byte) structuredRecord1.get(key),
+                                                  (Byte) structuredRecord2.get(key));
+    } else {
+      throw new IllegalArgumentException("Sorting can be performed only on the following CDAP data types:  " +
+                                           "String, Int, Long, Float, Double, Byte");
+    }
+    return res;
+  }
+
+  @Override
```
Comment (nitinmotgi) [inline orderby-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java:141]: Doc

File: orderby-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+  public void write(DataOutput dataOutput) throws IOException {
```
Comment (nitinmotgi) [inline orderby-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java:145]: Doc

File: orderby-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java (hunk 3 above)
Comment (nitinmotgi) [inline orderby-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java:80]: Take it back. You have implemented it below.

File: orderby-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+   * @return structured records sorted on the keys
+   */
+  private int compareNestedRecords(String key, String value, StructuredRecord structuredRecord1,
+                                   StructuredRecord structuredRecord2) {
+
+    if (key.contains(FIELD_DELIMITER)) {
+      String[] nestedRecord = key.split(FIELD_DELIMITER);
+      Schema.Field field = structuredRecord1.getSchema().getField(nestedRecord[0].trim());
+      Schema.Type type = field.getSchema().getType();
+      if (type.equals(Schema.Type.RECORD)) {
```
Comment (nitinmotgi) [inline orderby-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java:213]: Don't you want to handle UNION and such ?

File: orderby-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+                                    structuredRecord2);
+      } else {
+        throw new IllegalArgumentException("Nested structured defined by -> should be of type RECORD");
+      }
+    } else {
+      return compareRecords(key, value, structuredRecord1, structuredRecord2);
+    }
+  }
+
+  public String getStructureRecordJSON() {
```
Comment (nitinmotgi) [inline orderby-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java]: StructuredRecord not StructureRecord

File: orderby-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKeyDriver.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+public class CompositeKeyDriver {
+  public static void main(String[] args) throws Exception {
+    Job job = Job.getInstance(new Configuration(), "Order By");
+    job.setJarByClass(CompositeKeyDriver.class);
+
+    job.setMapperClass(CompositeKeyMapper.class);
+    job.setMapOutputKeyClass(CompositeKey.class);
+    job.setMapOutputValueClass(Text.class);
+
+    job.setPartitionerClass(CompositeKeyPartitioner.class);
```
Comment (nitinmotgi) [inline orderby-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKeyDriver.java:35]: @albertshau how can we get this into planner or app -- not sure which one. What do you think is the best approach to be able set the different comparator / partitioner when OrderBy is included. Please advice.

Comment (albertshau) [inline orderby-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKeyDriver.java:35]: what are we doing here? This is just a mapreduce job and not a plugin at all.

Anything we do needs to work for both Spark and for MapReduce. We should not be relying on such mapreduce specific features.

Can't order by be implemented by simply grouping with one partition (reducer)?

File: orderby-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java (hunk 6 above)
Comment (ananya-bhattacharya) [inline orderby-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java:213]: The if(#172) statement checks, if the path specified by user for selecting the nested field is valid, by checking the parent is of type RECORD.
For the current implementation we are only supporting simple types with nullable (as UNION): String, Int, Long, Short, Float, Double, Byte as mentioned in the use-case.

File: orderby-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKeyDriver.java (hunk 8 above)
Comment (ananya-bhattacharya) [inline orderby-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKeyDriver.java:35]: To sort the structured records using sort and group comparator we had used the MR job as a first part of the implementation of the Order-by plugin.
https://wiki.cask.co/display/CE/Order+By -  Implementation Tips #1 and #2
Please let me know, if this approach has to be changed and if we need to explore the plugin classes for implementing the order-by plugin.

**Comment by abhinavbansal246** [comment]
Closing PR against release, please open against develop

**Pull request closed without merging**